lib/ansible/plugins/connection/lxc.py

```diff
@@ -0,0 +1,210 @@
+        fcntl.fcntl(fd, fcntl.F_SETFD, fcntl.FD_CLOEXEC)
+        return fd
+
+    def exec_command(self, cmd, in_data=None, sudoable=False):
+        ''' run a command on the chroot '''
+        super(Connection, self).exec_command(cmd, in_data=in_data, sudoable=sudoable)
+
+        executable = C.DEFAULT_EXECUTABLE.split()[0] if C.DEFAULT_EXECUTABLE else '/bin/sh'
+        local_cmd = [executable, '-c', cmd]
+
+        read_stdout, write_stdout = os.pipe()
+        read_stderr, write_stderr = os.pipe()
+
+        kwargs = {
+            'stdout': self._set_nonblocking(write_stdout),
+            'stderr': self._set_nonblocking(write_stderr),
+            'env_policy': _lxc.LXC_ATTACH_CLEAR_ENV
+        }
+
+        if in_data:
+            read_stdin, write_stdin = os.pipe()
+            kwargs['stdin'] = self._set_nonblocking(read_stdin)
+        else:
+            read_stdin, write_stdin = None, None
+
+        display.vvv("EXEC %s" % (local_cmd), host=self.container_name)
+        pid = self.container.attach(_lxc.attach_run_command, local_cmd, **kwargs)
+        if pid == -1:
+            msg = "failed to attach to container %s" % self.container_name
+            raise errors.AnsibleError(msg)
+
+        os.close(write_stdout)
+        os.close(write_stderr)
+        if read_stdin:
+            os.close(read_stdin)
+
+        return self._communicate(pid,
+                                 in_data,
+                                 write_stdin,
+                                 read_stdout,
+                                 read_stderr)
+
+    def put_file(self, in_path, out_path):
+        ''' transfer a file from local to lxc '''
+        super(Connection, self).put_file(in_path, out_path)
+        display.vvv("PUT %s TO %s" % (in_path, out_path), host=self.container_name)
+
+        if not os.path.exists(in_path):
+            msg = "file or module does not exist: %s" % in_path
+            raise errors.AnsibleFileNotFound(msg)
+        try:
+            try:
+                src_file = open(in_path, "rb")
+            except IOError:
+                traceback.print_exc()
+                raise errors.AnsibleError("failed to open input file to %s" % in_path)
+
+            def write_file(args):
+                with open(out_path, 'wb+') as dst_file:
+                    shutil.copyfileobj(src_file, dst_file)
+            try:
+                self.container.attach_wait(write_file, None)
+            except IOError:
+                traceback.print_exc()
+                msg = "failed to transfer file to %s" % out_path
+                raise errors.AnsibleError(msg)
+        finally:
+            if src_file:
+                src_file.close()
+
+    def fetch_file(self, in_path, out_path):
+        ''' fetch a file from lxc to local '''
+        super(Connection, self).fetch_file(in_path, out_path)
+        display.vvv("FETCH %s TO %s" % (in_path, out_path), host=self.container_name)
+
+        if not os.path.exists(in_path):
```

Comment:
How does this work?  in_path should a path relative to the root inside the container.  If we check for its existence here we end up checking whether it exists on the host, don't we?